api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import org.apache.iceberg.Table;
+
+/**
+ * An API that should be implemented by query engine integrations for providing actions.
+ */
+public interface ActionsProvider {
```

Comment:
One thing that I would be concerned about here is this interface grows forever with every new action we add. I can imagine some actions are going to be engine (or a subgroup of engines) specific as well. I wonder if its possible to have a more dynamic way of registering actions (eg an action catalog of sorts) or some categories and an interface per category?

Reply:
Could you elaborate a bit more on the idea, @rymurr? Would we load actions by name?

Reply:
@aokolnychyi I think we could do either by name, enum or even by interface eg `actionCatalog.action(RewriteDataFiles.class)`. Presumably the catalog would need to expose a list of actions it supports as well. Not sure what others think on this idea? If configurations in configurable actions are engine configurations we may be able to add that to the catalog interface rather than each action?

Reply:
I am open to considering this idea but I am not sure it is going to help.

It sounds like our `SparkProcedures` class. Loading by name/enum/class, will most likely give us a parent action interface, not a specific interface to call. Also, we will probably need to change how arguments are passed. And each time we add a new action, we will have to add a new constant to the enum, which is more or less equivalent to adding a new method here with a default implementation.

I see this API kind of close to our `Table` API. That one also has a number of methods but it is pretty convenient. That said, I do get the concern of having a lot of methods in this API but we can address it differently. For example, we will support multiple compaction strategies. Instead of offering a separate action for each type, we can make the configuration flexible.

Reply:
I agree with Anton here. If we were to load by name, we wouldn't get a specific interface to configure the action. We could get the interface if we loaded by passing in the interface, but then I don't see how we'd be much better off. The API would be awkward (you have to find and supply the action interface you want) and the benefit is avoiding a few methods here.

I think what Anton has in the PR is a good step forward instead.

Reply:
agreed :-)

Reply:
Let me resolve this one to simplify the review. Thanks, @rymurr @rdblue!